src/index.ts

```diff
@@ -18,6 +19,142 @@ export interface Attachment {
+        this._timestamp = timestamp
+        this._videoTimestamp = videoTimestamp
+        this.properties = { property: properties }
+        this.code = code
+        if (status === Status.Failed) {
+            this.failure = { output }
+        }
+        if (status === Status.Skipped) {
+            this.skipped = { output }
+        }
+    }
+}
+
+/**
+ * JUnitTestSuite represents a testsuite in JUnit format.
+ */
+export class JUnitTestSuite {
+    _name: string
+    _status: Status
+    _tests: number
+    _failures: number
+    _skipped: number
+    _time: number
+    properties: object
+    testcase: JUnitTestCase[] // to correctly build an array using XMLBuilder, testcase should be kept in the singular form.
+
+    constructor(
+        name: string,
+        status: Status,    
+        properties: object[],
+        testcases: JUnitTestCase[],
+    ) {
+        this._name = name
+        this._status = status
+        this.properties = { property: properties }
+        this.testcase = testcases
+        this._tests = testcases.length
+        this._failures = 0
+        this._skipped = 0
+        this._time = 0
+        this.testcase.forEach(testcase => {
+            if (testcase._status === Status.Failed) {
+                this._failures += 1
+            }
+            if (testcase._status === Status.Skipped) {
+                this._skipped += 1
+            }
+            this._time += testcase._time
+        })
+    }
+}
+
+/**
+ * JUnitReport represents a JUnit report.
+ */
+export class JUnitReport {
+    _status: Status
+    _tests: number
+    _failures: number
+    _skipped: number
+    _time: number
+    testsuite: JUnitTestSuite[] // to correctly build an array using XMLBuilder, testsuite should be kept in the singular form.
+    properties: object
+
+    constructor(
+        testsuites: JUnitTestSuite[],
+        status: Status,
+        properties: object[],
+    ) {
+        this.testsuite = testsuites
+        this._status = status
+        this.properties = { property: properties }
+        this._failures = 0
+        this._skipped = 0
+        this._tests = 0
+        this._time = 0
+        this.testsuite.forEach(testsuite => {
+            this._failures += testsuite._failures
+            this._skipped += testsuite._skipped
+            this._tests += testsuite._tests
+            this._time += testsuite._time
+        })
+    }
+}
+
+/**
+ * toProperty converts attachments and metadata to JUnit properties.
+ */
+function toProperty(
+    attachments: Attachment[],
+    metadata: object,
+) {
+    const properties: object[] = [];
+    attachments.forEach(attachment => {
+        properties.push({
+            _name: 'attachment',
+            _value: attachment.name,
+            '#text': attachment.path,
+        })
+    })
+    Object.keys(metadata).forEach(key => {
+        properties.push({
+            _name: key,
+            _value: (metadata as any)[key],
+        })
+    });
```

Comment:
Since you need both key & value, wouldn't this be more natural?
```suggestion
    Object.entries(metadata).forEach(([key, value]) => {
        properties.push({
            _name: key,
            _value: value,
        })
    });
```

Reply:
One thing we also need to think about and deal with, is if the value of the metadata is also another object. Don't frameworks like TestCafe allow arbitrary objects as metadata?

Just want to make sure that we have proper safeguards, if that's what we need, around user provided data and making sure that our report doesn't break.

Reply:
I've tried with objects. That output would be like
```
<property name="items">
      <_value>
        <test1>1</test1>
        <test2>2</test2>
      </_value>
</property>
```
Not canonical JUnit format but meets XML naming convention https://www.w3schools.com/xml/xml_elements.asp
```
    Element names are case-sensitive
    Element names must start with a letter or underscore
    Element names cannot start with the letters xml (or XML, or Xml, etc)
    Element names can contain letters, digits, hyphens, underscores, and periods
    Element names cannot contain spaces
```